File: frontend/src/lib/formulaRegistry.ts
/**
 * PATH: frontend/src/lib/formulaRegistry.ts
 * PURPOSE: Load sealed formula-registry.json — every investor number ID + cite.
 */
import registryJson from "@/fixtures/formula-registry.json"

export type FormulaStatus = "audited" | "documented" | "pending"

export type FormulaEntry = {
  id: string
  name: string
  expression: string
  surfaces: string[]
  compute: string[]
  display: string[]
  audit: string[]
  reference: { kind: string; cite: string }
  status: FormulaStatus
}

export type FormulaRegistry = {
  schema_version: number
  title: string
  authority: string
  formulas: FormulaEntry[]
}

export const formulaRegistry = registryJson as FormulaRegistry

export function formulaById(id: string): FormulaEntry {
  const row = formulaRegistry.formulas.find((f) => f.id === id)
  if (!row) throw new Error(`Unknown formula id: ${id}`)
  return row
}

/** Tip suffix: `[F_VS_MEDIAN_PCT]` for auditability in the UI. */
export function formulaTip(id: string, body: string): string {
  const row = formulaById(id)
  return `${body} [${row.id}: ${row.expression}]`
}

export function auditedFormulaIds(): string[] {
  return formulaRegistry.formulas.filter((f) => f.status === "audited").map((f) => f.id)
}

File: backend/app/contracts/formula_registry.py
"""
PATH: backend/app/contracts/formula_registry.py
PURPOSE: Load contracts/formula-registry.json — provenance for every investor number.
"""

from __future__ import annotations

import json
from functools import lru_cache
from typing import Any

from app.contracts.paths import contracts_dir


@lru_cache(maxsize=1)
def load_formula_registry() -> dict[str, Any]:
    path = contracts_dir() / "formula-registry.json"
    return json.loads(path.read_text(encoding="utf-8"))


def formula_ids() -> list[str]:
    return [f["id"] for f in load_formula_registry()["formulas"]]


def formula_by_id(formula_id: str) -> dict[str, Any]:
    for row in load_formula_registry()["formulas"]:
        if row["id"] == formula_id:
            return row
    raise KeyError(formula_id)


def audited_formula_ids() -> list[str]:
    return [f["id"] for f in load_formula_registry()["formulas"] if f.get("status") == "audited"]
